pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -101,48 +102,113 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, upd
 
 	var deletedClusters []string
 	var services []*model.Service
-	// holds clusters per service, keyed by hostname.
+	// Holds clusters per service, keyed by hostname.
 	serviceClusters := make(map[string]sets.String)
-	// holds service ports, keyed by hostname.
-	// inner map holds port and its cluster name.
-	servicePorts := make(map[string]map[int]string)
+	// Holds service ports, keyed by hostname.Inner map port and its cluster name.
+	// This is mainly used when service is updated and a port has been removed.
+	servicePortClusters := make(map[string]map[int]string)
+	// Holds subset clusters per service, keyed by hostname.
+	subsetClusters := make(map[string]sets.String)
 
 	for _, cluster := range watched.ResourceNames {
 		// WatchedResources.ResourceNames will contain the names of the clusters it is subscribed to. We can
-		// check with the name of our service (cluster names are in the format outbound|<port>||<hostname>).
-		_, _, svcHost, port := model.ParseSubsetKey(cluster)
-		sets.InsertOrNew(serviceClusters, string(svcHost), cluster)
-		if servicePorts[string(svcHost)] == nil {
-			servicePorts[string(svcHost)] = make(map[int]string)
+		// check with the name of our service (cluster names are in the format outbound|<port>|<subset>|<hostname>).
+		_, subset, svcHost, port := model.ParseSubsetKey(cluster)
+		if subset == "" {
+			sets.InsertOrNew(serviceClusters, string(svcHost), cluster)
+		} else {
+			sets.InsertOrNew(subsetClusters, string(svcHost), cluster)
+		}
+		if servicePortClusters[string(svcHost)] == nil {
+			servicePortClusters[string(svcHost)] = make(map[int]string)
 		}
-		servicePorts[string(svcHost)][port] = cluster
+		servicePortClusters[string(svcHost)][port] = cluster
 	}
 
-	// In delta, we only care about the services that have changed.
 	for key := range updates.ConfigsUpdated {
-		// get the service that has changed.
-		service := updates.Push.ServiceForHostname(proxy, host.Name(key.Name))
-		// if this service removed, we can conclude that it is a removed cluster.
-		if service == nil {
-			for cluster := range serviceClusters[key.Name] {
+		switch key.Kind {
+		case kind.ServiceEntry:
+			services, deletedClusters = configgen.deltaFromServices(key, proxy, updates.Push, serviceClusters,
+				servicePortClusters, subsetClusters)
+		case kind.DestinationRule:
+			services, deletedClusters = configgen.deltaFromDestinationRules(key, proxy, updates.Push, subsetClusters)
+		}
+	}
+	if services == nil {
+		// No matching services, just rebuild with all services.
+		services = proxy.SidecarScope.Services()
```

Comment:
If servuces == nil, why does not it mean no update needed

Reply:
It is. Just added for safety for now so that if our logic has any issues, we can cover it up. We can remove it. Let me know if you would like to remove it.

Reply:
Yeah, i want to remove, then we can verify if it works or not